File: src/main/java/org/parboiled/parserunners/TracingParseRunner.java
/*
 * Copyright (C) 2009-2010 Mathias Doenitz
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.parboiled.parserunners;

import org.jetbrains.annotations.NotNull;
import org.parboiled.Context;
import org.parboiled.MatchHandler;
import org.parboiled.MatcherContext;
import org.parboiled.Rule;
import org.parboiled.buffers.InputBuffer;
import org.parboiled.common.Predicate;
import org.parboiled.common.Predicates;
import org.parboiled.common.Tuple2;
import org.parboiled.matchers.Matcher;
import org.parboiled.support.MatcherPath;

/**
 * A {@link org.parboiled.parserunners.ParseRunner} implementation used for debugging purposes.
 * It exhibits the same behavior as the {@link ReportingParseRunner} but collects debugging information as to which
 * rules did match and which didn't.
 */
public class TracingParseRunner<V> extends BasicParseRunner<V> {

    private final StringBuilder log = new StringBuilder();
    private final Predicate<Tuple2<Context<?>, Boolean>> filter;

    /**
     * Creates a new TracingParseRunner instance for the given rule.
     *
     * @param rule the parser rule
     */
    public TracingParseRunner(@NotNull Rule rule) {
        this(rule, Predicates.alwaysTrue());
    }

    /**
     * Creates a new TracingParseRunner instance for the given rule.
     * The given filter is used to select the matchers to print tracing statements for.
     * NOTE: The given filter must be of type Predicate<Tuple2<Context<?>, Boolean>>. The reason this type is not
     * directly specified in the constructors signature is that this would make predicate expressions using the
     * {@link Predicates} operations and the predefined predicate constructors in {@link org.parboiled.support.Filters}
     * much more cumbersome to write (due to Java limited type parameters inference logic you would have to explicitly
     * state the type parameters in many places).
     *
     * @param rule   the parser rule
     * @param filter the matcher filter selecting the matchers to print tracing statements for. Must be of type
     *               Predicate<Tuple2<Context<?>, Boolean>>.
     */
    @SuppressWarnings({"unchecked"})
    public TracingParseRunner(@NotNull Rule rule, @NotNull Predicate<?> filter) {
        super(rule);
        this.filter = (Predicate<Tuple2<Context<?>, Boolean>>) filter;
    }

    /**
     * Retrieves a string containing all generated log messages.
     *
     * @return the log messages
     */
    public String getLog() {
        return log.toString();
    }

    @SuppressWarnings({"SimplifiableIfStatement"})
    @Override
    protected boolean runRootContext() {
        // run a basic match
        if (runRootContext(new Handler(log, filter), true)) {
            return true;
        }

        // ok, we have a parse error, so run again without fast string matching and with our recording handler
        RecordingParseRunner.Handler recordingHandler = new RecordingParseRunner.Handler();
        if (runRootContext(recordingHandler, false)) {
            throw new IllegalStateException(); // we failed before so we should really be failing again
        }

        // finally perform a third, reporting run (now that we know the error location)
        ReportingParseRunner.Handler reportingHandler = new ReportingParseRunner.Handler(
                recordingHandler.getErrorIndex());
        if (runRootContext(reportingHandler, false)) {
            throw new IllegalStateException(); // we failed before so we should really be failing again
        }

        return false;
    }

    /**
     * A {@link org.parboiled.MatchHandler} implementation that reports the {@link org.parboiled.errors.InvalidInputError} at a given error index.
     * For the actual matching this handler relies on another, inner {@link org.parboiled.MatchHandler} instance it delegates to.
     */
    public static class Handler implements MatchHandler {
        private final StringBuilder log;
        private final Predicate<Tuple2<Context<?>, Boolean>> filter;
        private MatcherPath lastPath;

        public Handler(StringBuilder log, Predicate<Tuple2<Context<?>, Boolean>> filter) {
            this.log = log;
            this.filter = filter;
        }

        public boolean matchRoot(MatcherContext<?> rootContext) {
            log.setLength(0);
            lastPath = null;
            return rootContext.runMatcher();
        }

        @SuppressWarnings({"unchecked"})
        public boolean match(MatcherContext<?> context) {
            Matcher matcher = context.getMatcher();
            boolean matched = matcher.match(context);
            if (filter.apply(new Tuple2<Context<?>, Boolean>(context, matched))) {
                print(context, matched);
            }
            return matched;
        }

        private void print(MatcherContext<?> context, boolean matched) {
            InputBuffer.Position pos = context.getInputBuffer().getPosition(context.getCurrentIndex());
            MatcherPath path = context.getPath();
            MatcherPath prefix = lastPath != null ? path.commonPrefix(lastPath) : null;
            if (prefix != null && prefix.length() > 1) log.append("..(").append(prefix.length() - 1).append(")../");
            log.append(path.toString(prefix != null ? prefix.parent : null));
            String line = context.getInputBuffer().extractLine(pos.line);
            log.append(", ")
                    .append(matched ? "matched" : "failed")
                    .append(", cursor at ")
                    .append(pos.line)
                    .append(':')
                    .append(pos.column)
                    .append(" after \"")
                    .append(line.substring(0, Math.min(line.length(), pos.column - 1)))
                    .append("\"\n");
            lastPath = path;
        }

    }

}

File: src/main/scala/org/parboiled/scala/parserunners/TracingParseRunner.scala
package org.parboiled.scala.parserunners

import org.parboiled.parserunners.{TracingParseRunner => PTracingParseRunner}
import org.parboiled.scala._
import org.parboiled.Context
import utils.Predicate
import org.parboiled.common.{Predicate => PPredicate, Tuple2 => T2}
import org.parboiled.support.MatcherPath

/**
 * A wrapper for org.parboiled.parserunners.TracingParseRunner which returns a scala ParsingResult.
 * It provides for the ability to attach filter expressions for limiting the tracing printout to certain input and/or
 * grammar areas.
 * Note that the ParseRunner only accepts rules with zero or one value type parameter, as parsers leaving more
 * than one value on the value stack are considered to be bad style.
 */
object TracingParseRunner {
  def apply(rule: Rule0) = new TracingParseRunner[Nothing](new PTracingParseRunner[Nothing](rule))

  def apply[V](rule: Rule1[V]) = new TracingParseRunner[V](new PTracingParseRunner[V](rule))
}

class TracingParseRunner[V](val inner: PTracingParseRunner[V]) extends ParseRunner[V] {
  def run(input: Input): ParsingResult[V] = ParsingResult(inner.run(input.inputBuffer))

  /**
   * Returns a new TracingParseRunner with the attached filter predicate limiting the tracing printout.
   * You can supply an expression consisting of either the predefined Predicates (see the Lines, Rules, Matched and
   * Mismatched objects) or custom functions with one of the following signatures as TracingPredicates:
   * Context[_] => Boolean
   * (Context[_], Boolean) => Boolean
   *
   * For example: "Lines(10 until 20) && Rules.below(parser.Factor)".
   */
  def filter(filter: TracingPredicate) =
    new TracingParseRunner[V](new PTracingParseRunner[V](inner.rootMatcher, new PPredicate[T2[Context[_], Boolean]] {
      def apply(t: T2[Context[_], Boolean]) = filter(t.a, t.b)
    }))

  def traceLog: String = inner.getLog
}

abstract class TracingPredicate extends Predicate[(Context[_], Boolean)]

object TracingPredicate {
  implicit def fromRawPredicate(p: Predicate[(Context[_], Boolean)]): TracingPredicate = new TracingPredicate {
    def apply(t: (Context[_], Boolean)) = p(t)
  }

  implicit def fromContextPredicate(f: Context[_] => Boolean): TracingPredicate = new TracingPredicate {
    def apply(t: (Context[_], Boolean)) = f(t._1)
  }
}

object Lines {
  def from(firstLine: Int): TracingPredicate = toPredicate(_ >= firstLine)

  def until(lastLine: Int): TracingPredicate = toPredicate(_ <= lastLine)

  def apply(lineRange: Range): TracingPredicate = {
    require(lineRange.step == 1, "Ranges with step != 1 are not supported here")
    toPredicate(line => { lineRange.start <= line && line < lineRange.end })
  }

  private def toPredicate(f: Int => Boolean): TracingPredicate =
    (c: Context[_]) => f(c.getInputBuffer.getPosition(c.getCurrentIndex).line)
}

object Rules {
  def only(rules: Rule*): TracingPredicate = {
    val matchers = rules.map(_.matcher)
    (c:Context[_]) => matchers.contains(c.getMatcher)
  }

  def below(rules: Rule*): TracingPredicate = {
    val matchers = rules.map(_.matcher)
    toPredicate(path => { matchers.exists(m => (path.element.matcher ne m) && path.contains(m)) })
  }

  def apply(rules: Rule*): TracingPredicate = {
    val matchers = rules.map(_.matcher)
    toPredicate(path => { matchers.exists(path.contains) })
  }

  private def toPredicate(f: MatcherPath => Boolean): TracingPredicate = (c: Context[_]) => f(c.getPath)
}

object Matched extends TracingPredicate {
  def apply(t: (Context[_], Boolean)) = t._2
}

object Mismatched extends TracingPredicate {
  def apply(t: (Context[_], Boolean)) = !t._2
}